File: Top_100_Liked_Questions/15._3Sum/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let ans = [];
    let found = {};
    
    for (let i=0; i<nums.length; i++) {
        let seen = {};
        for (let j=i+1; j<nums.length; j++) {
            let complement = -nums[i] - nums[j];
            if (seen.hasOwnProperty(complement)) {
                mx = Math.max(nums[i], nums[j], complement);
                mn = Math.min(nums[i], nums[j], complement);
                if (!found.hasOwnProperty([mn, mx])) {
                    ans.push([nums[i], nums[j], complement]);
                    found[[mn, mx]] = 0;
                }
            }
            seen[nums[j]] = 0;
        }
    }
    
    return ans;
};

File: Top_100_Liked_Questions/15._3Sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # choose two and check if complement exists in the hash
        ans = []
        found = set()
        for i in range(len(nums)):
            seen = set()
            for j in range(i+1, len(nums)):
                complement = -nums[i] - nums[j]
                if complement in seen:
                    mx = max(nums[i], nums[j], complement)
                    mn = min(nums[i], nums[j], complement)
                    if (mn, mx) not in found:
                        ans.append([nums[i], nums[j], complement])
                        found.add((mn, mx))
                
                seen.add(nums[j])
        
        return ans